src/azul/project/hca/transformers.py

```diff
@@ -249,22 +254,46 @@ def create_documents(self,
                             manifest=manifest,
                             metadata_files=metadata_files)
 
+        bundle_uuid = str(bundle.uuid)
+        simplified_project = _project_dict(bundle)
+
         # Create ElasticSearch documents
         for project in bundle.projects.values():
-            visitor = TransformerVisitor()
-
-            # Visit the relatives of file
-            # project.accept(visitor)  # Visit descendants
-            # project.ancestors(visitor)
+            contents = dict(specimens=self._get_unique_list_of_specimens(bundle),
+                            files=self._get_unique_list_of_files(bundle),
+                            processes=self._get_unique_list_of_process(bundle),
+                            project=simplified_project)
 
-            # Assign the contents to the ES doc
-            contents = dict(specimens=_specimen_dict(visitor.specimens),
-                            files=list(visitor.files.values()),
-                            processes=list(visitor.processes.values()),
-                            project=_project_dict(bundle))
             es_document = ElasticSearchDocument(entity_type=self.entity_name,
                                                 entity_id=str(project.document_id),
-                                                bundles=[Bundle(uuid=str(bundle.uuid),
+                                                bundles=[Bundle(uuid=bundle_uuid,
                                                                 version=bundle.version,
                                                                 contents=contents)])
             yield es_document
+
+    def _get_unique_list_of_specimens(self, bundle):
+        return _specimen_dict({
+            specimen.document_id: specimen
+            for specimen in bundle.specimens
+        }),
+
+    def _get_unique_list_of_files(self, bundle):
+        return [
+            TransformerVisitor.file_dict(file)
+            for file in bundle.files.values()
+        ]
+
+    def _get_unique_list_of_process(self, bundle):
+        process_map = dict()
+
+        for process in bundle.processes.values():
+            if not process.protocols:
+                continue
+
+            process_map.update({
+                f'PC{process.document_id}-PL{protocol.document_id}': TransformerVisitor.merge_process_protocol(process,
```

Comment:
nit: do you think you could make this line shorter? The line break after process looks weird.  Maybe break after the key?

Reply:
We could make this shorter but we have to write the normal `for` loop. :)